### PR TITLE
chore(rust): add ockam status command

### LINKED ISSUE #3460: Implement `ockam status` command
Create a new command `ockam status` at the root of the [`ockam_command`](https://github.com/build-trust/ockam/tree/develop/implementations/rust/ockam/ockam_command/src) crate that will show the identities that have been enrolled.

An identity is enrolled when running the `ockam enroll` command, and currently, it doesn't support passing a specific node, so it's able to enroll only the `default_identity`. This is a limitation on the `ockam enroll` command, so the `ockam status` should be generic and support showing the status of multiple identities.

This information can be written on the config file that is available on the `CommandGlobalOpts` (more info [here](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_command/src/lib.rs#L238)).

A possible solution would be to write into that file after the enroll has happened, [before exiting the command](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_command/src/enroll.rs#L58).

This command could be further extended in another PR so it also shows a short version of `ockam node list`.

---

_We love helping new contributors!_
_If you have questions or need help as you work on your first Ockam contribution, please leave a comment on [this discussion](https://github.com/ockam-network/ockam/discussions/1081)._
_If you're looking for other issues to contribute to, checkout [this discussion](https://github.com/build-trust/ockam/discussions/3206) and labels -  https://github.com/build-trust/ockam/labels/good%20first%20issue or https://github.com/build-trust/ockam/labels/help%20wanted_

### PR DESCRIPTION
**Pull request by hargut**
## Current Behavior
No `status` command available.

## Proposed Changes
Introduce `status` command, providing information about enrolled or all active identities and linked nodes.

Solves https://github.com/build-trust/ockam/issues/3460

## Checks
- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

### CODE REVIEW

**Comment by hargut** [comment]
Example output from  `ockam status`:
```
Identity[0]
  Default: yes
  Name: da80759c
  State Path: /home/hargut/.ockam/identities/da80759c.json
  Config Identifier: P36cae565dfde8d83f83d89d63d91a175ede4f6ea6bf799fc4ccfe8a4c8aa29b3
  Linked Nodes:
    Node[0]:
      Name: 0afffc7d
      Status: Stopped
```

**Comment by adrianbenavides** [comment]
Thanks @hargut for submitting this PR. From what I understood reading the [issue](https://github.com/build-trust/ockam/issues/3460), this command should show the identities that were enrolled with `ockam enroll`, and not just listing the available identities.

To do that, you should modify the `IdentityConfig` (the information we store about identities on disk) with a new field to track whether an identity was enrolled. This information should be filled after an identity has been successfully enrolled (`enroll` command). Then, on the `IdentitiesState`, you could add a method to retrieve the identities that were enrolled. For now, you can keep the shown information simple (i.e. no need to show the linked nodes). 

Let me know if you have any questions or need help with anything!

**Comment by hargut** [comment]
Hi @adrianbenavides,

I've added the part to store the enrollment status with the identity, and modified the `status` command to only display the information for enrolled identities. With the `--all` flag details will be printed for all identities including non-enrolled.
```
Identity[0]
  Default: yes
  Name: f6946397
  State Path: /home/hargut/.ockam/identities/f6946397.json
  Config Identifier: P47018045bdfeeb1c55158d7725ebdda8cabb8374efc4db3c3b08a791e77113b4
  Enrollment Status:
    Enrolled: yes
    Timestamp: 1671975757
  Linked Nodes:
    Node[0]:
      Name: 9f82ddc9
      Status: Running
```

File: implementations/rust/ockam/ockam_api/src/cli_state.rs [hunk 1]
```diff
@@ -322,6 +373,29 @@ impl PartialEq for IdentityConfig {
...
+impl Display for EnrollmentStatus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if self.is_enrolled {
+            writeln!(f, "Enrolled: yes")?;
+        } else {
+            writeln!(f, "Enrolled: no")?;
+        }
+
+        match self.created_at.duration_since(UNIX_EPOCH) {
+            Ok(time) => writeln!(f, "Timestamp: {}", time.as_secs())?,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/cli_state.rs:400]: I think it would be best to show a human-readable format here. I'd suggest adding the [`time`](https://github.com/time-rs/time) dependency and format the `SystemTime` to [Iso8601](https://docs.rs/time/latest/time/format_description/well_known/struct.Iso8601.html#)

File: implementations/rust/ockam/ockam_command/src/enroll.rs [hunk 2]
```diff
@@ -357,3 +360,27 @@ impl Auth0Service {
...
     }
 }
+
+async fn update_enrolled_identity(
+    ctx: &Context,
+    opts: &CommandGlobalOpts,
+    node_name: &str,
+) -> Result<()> {
+    let mut identities = opts.state.identities.list()?;
+    identities.push(opts.state.identities.default()?);
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/enroll.rs]: Why is this needed? Shouldn't the default identity should be returned by `list()`?

File: implementations/rust/ockam/ockam_command/src/status.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            println!("{:2}Default: yes", "")
+        }
+        for line in identity.to_string().lines() {
+            println!("{:2}{}", "", line);
+        }
+        if !node_details.is_empty() {
+            println!("{:2}Linked Nodes:", "");
+            for (n_idx, node) in node_details.iter().enumerate() {
+                println!("{:4}Node[{}]:", "", n_idx);
+                if node.identity.identifier() == &identity.config.identifier {
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/status.rs]: Note that the two prints done before this line will be printed for every identity, even if only one identity has linked nodes. You should buffer the print contents and print it after you know for sure that the current identity has linked nodes.

File: implementations/rust/ockam/ockam_command/src/enroll.rs [hunk 4]
```diff
@@ -357,3 +360,27 @@ impl Auth0Service {
...
+
+    let node_state = opts.state.nodes.get(node_name)?;
+    let node_identity = node_state.config.identity(ctx).await?;
+
+    for mut identity in identities {
+        if node_identity.identifier() == &identity.config.identifier {
+            identity.config.enrollment_status = Some(EnrollmentStatus {
+                is_enrolled: true,
+                created_at: SystemTime::now(),
+            });
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/enroll.rs]: Since the `EnrollmentStatus` variables are constants, you could add a constructor so you can write this line like `identity.config.enrollment_status = Some(EnrollmentStatus::enrolled());`.

**Review by adrianbenavides — CHANGES_REQUESTED** [review]
Looks pretty great! I've left some suggestions, thanks! 👍

File: implementations/rust/ockam/ockam_api/src/cli_state.rs (hunk 1 above)
Comment (hargut) [reply implementations/rust/ockam/ockam_api/src/cli_state.rs:400]: I've upfront already considered and tested that approach, but it lead to an issue within the no_std check in the rust workflow. Therefore I've for now just included the POSIX style output. So far I've not tested the chrono crate which might be another solution.

https://github.com/hargut/ockam/blob/feature/add-status-command/.github/workflows/rust.yml#L159
https://github.com/hargut/ockam/actions/runs/3776101474/jobs/6419173662#step:7:65

https://github.com/chronotope/chrono/blob/378efb157d674c01761f538d4450705c2b1766a4/.github/workflows/test.yml#L87

File: implementations/rust/ockam/ockam_command/src/enroll.rs (hunk 2 above)
Comment (hargut) [reply implementations/rust/ockam/ockam_command/src/enroll.rs]: Thank you for pointing that out, I've checked it again and it is not needed.

The locations where two different directories so I had assumed it is two different files, but the default identity is just a symlink to the identities directory.

File: implementations/rust/ockam/ockam_command/src/status.rs (hunk 3 above)
Comment (hargut) [reply implementations/rust/ockam/ockam_command/src/status.rs]: Good point, I'll adjust that part.

File: implementations/rust/ockam/ockam_command/src/enroll.rs (hunk 4 above)
Comment (hargut) [reply implementations/rust/ockam/ockam_command/src/enroll.rs]: Will adjust that. :+1:

**Comment by hargut** [comment]
Hi @adrianbenavides,

thank you for your review. Happy New Year!

Best regards,
Harald

File: implementations/rust/ockam/ockam_api/src/cli_state.rs (hunk 1 above)
Comment (hargut) [reply implementations/rust/ockam/ockam_api/src/cli_state.rs:400]: @adrianbenavides
Please let me know your thoughts on how to handle that point.

Comment (adrianbenavides) [reply implementations/rust/ockam/ockam_api/src/cli_state.rs:400]: OK, let's not worry about this. I'll open a new issue to tackle this in a separate PR.

**Review by adrianbenavides — DISMISSED** [review]
Thanks again @hargut and happy new year!
LGTM 🚀

**Review by adrianbenavides — APPROVED** [review]
